plasma_cash/utils/merkle/sparse_merkle_tree.py

```diff
@@ -1,41 +1,52 @@
 from collections import OrderedDict
+
 from ethereum.utils import sha3
 
 
 class SparseMerkleTree(object):
 
-    def __init__(self, leaves={}):
-        self.depth = 256
+    def __init__(self, depth=256, leaves={}):
+        self.depth = depth
+        if len(leaves) > 2**(depth-1):
+            raise self.TreeSizeExceededException(
+                'tree with depth {} could not have {} leaves'.format(depth, len(leaves))
+            )
+
         # Sort the transaction dict by index.
         self.leaves = OrderedDict(sorted(leaves.items(), key=lambda t: t[0]))
-        self.default_nodes = [b'\x00' * 32]
-        self.create_default_nodes()
+        self.default_nodes = self.create_default_nodes(self.depth)
         if leaves:
-            self.root = self.create_tree()
+            self.root = self.create_tree(self.leaves, self.depth, self.default_nodes)
         else:
             self.root = self.default_nodes[self.depth - 1]
 
-    def create_default_nodes(self):
+    def create_default_nodes(self, depth):
         # Default nodes are the nodes whose children are both empty nodes at each level.
-        for level in range(1, self.depth):
-            prev_default = self.default_nodes[level - 1]
-            self.default_nodes.append(sha3(prev_default + prev_default))
+        default_nodes = [b'\x00' * 32]
+        for level in range(1, depth):
+            prev_default = default_nodes[level - 1]
+            default_nodes.append(sha3(prev_default + prev_default))
+        return default_nodes
 
-    def create_tree(self):
-        tree_level = self.leaves
-        for level in range(self.depth - 1):
+    def create_tree(self, ordered_leaves, depth, default_nodes):
```

Comment:
Just curious.
Why pass in `default_nodes` instead of using `self.default_nodes`?

Reply:
originally I had a plan to test all functions inside separately. In other words, there would be tests such as test_constructor, test_create_tree... It will be easier to do so. At the end, I failed on doing so but I feel it is better to inject dependency if it is not hard to do. So we don't need to worry about what side effect this method would bring and just focus on input/output.